File: packages/database/src/observability.ts
export type OperationalLogLevel = "debug" | "info" | "warn" | "error";
export type OperationalOutcome = "succeeded" | "failed" | "cancelled" | "idle";

export type OperationalLog = {
  level: OperationalLogLevel;
  service: "web" | "worker" | "migration" | "test";
  operation: string;
  correlationId: string;
  outcome: OperationalOutcome;
  eventId?: string;
  jobId?: string;
  handlerVersion?: number;
  durationMs?: number;
  safeErrorCode?: string;
};

export type OperationalLogRecord = {
  timestamp: string;
  level: OperationalLogLevel;
  service: OperationalLog["service"];
  operation: string;
  correlation_id: string;
  outcome: OperationalOutcome;
  event_id?: string;
  job_id?: string;
  handler_version?: number;
  duration_ms?: number;
  safe_error_code?: string;
};

const SAFE_TOKEN = /^[a-z0-9_.-]{1,120}$/;
const SAFE_IDENTIFIER = /^[A-Za-z0-9_.:-]{1,160}$/;

export function createOperationalLogRecord(
  log: OperationalLog,
  now = new Date(),
): OperationalLogRecord {
  if (!SAFE_TOKEN.test(log.operation))
    throw new TypeError("operation must be a bounded safe token");
  if (log.safeErrorCode && !SAFE_TOKEN.test(log.safeErrorCode))
    throw new TypeError("safe error code must be a bounded safe token");
  for (const [name, value] of [
    ["correlation ID", log.correlationId],
    ["event ID", log.eventId],
    ["job ID", log.jobId],
  ] as const) {
    if (value !== undefined && !SAFE_IDENTIFIER.test(value))
      throw new TypeError(`${name} must be a bounded safe identifier`);
  }
  if (
    log.durationMs !== undefined &&
    (!Number.isFinite(log.durationMs) || log.durationMs < 0)
  )
    throw new TypeError("duration must be a finite non-negative number");
  return {
    timestamp: now.toISOString(),
    level: log.level,
    service: log.service,
    operation: log.operation,
    correlation_id: log.correlationId,
    outcome: log.outcome,
    ...(log.eventId ? { event_id: log.eventId } : {}),
    ...(log.jobId ? { job_id: log.jobId } : {}),
    ...(log.handlerVersion !== undefined
      ? { handler_version: log.handlerVersion }
      : {}),
    ...(log.durationMs !== undefined ? { duration_ms: log.durationMs } : {}),
    ...(log.safeErrorCode ? { safe_error_code: log.safeErrorCode } : {}),
  };
}

export function emitOperationalLog(
  log: OperationalLog,
  sink: (line: string) => void = console.log,
): void {
  sink(JSON.stringify(createOperationalLogRecord(log)));
}

File: services/worker/src/cumulore_worker/observability.py
"""Small vendor-neutral structured logging boundary with an explicit safe schema."""

from __future__ import annotations

import json
from collections.abc import Mapping
from datetime import UTC, datetime
from typing import Literal

LogLevel = Literal["debug", "info", "warn", "error"]
LogOutcome = Literal["succeeded", "failed", "cancelled", "idle"]


def emit_operational_log(
    *,
    level: LogLevel,
    operation: str,
    correlation_id: str,
    outcome: LogOutcome,
    duration_ms: float,
    event_id: str | None = None,
    job_id: str | None = None,
    handler_version: int | None = None,
    safe_error_code: str | None = None,
    metrics: Mapping[str, int | float] | None = None,
) -> None:
    record: dict[str, object] = {
        "timestamp": datetime.now(UTC).isoformat(),
        "level": level,
        "service": "worker",
        "operation": _safe_token(operation),
        "correlation_id": _safe_identifier(correlation_id),
        "outcome": outcome,
        "duration_ms": max(0.0, round(duration_ms, 3)),
    }
    if event_id is not None:
        record["event_id"] = _safe_identifier(event_id)
    if job_id is not None:
        record["job_id"] = _safe_identifier(job_id)
    if handler_version is not None:
        record["handler_version"] = handler_version
    if safe_error_code is not None:
        record["safe_error_code"] = _safe_token(safe_error_code)
    if metrics is not None:
        record["metrics"] = {
            _safe_token(key): value for key, value in sorted(metrics.items())
        }
    print(json.dumps(record, separators=(",", ":"), sort_keys=True), flush=True)


def _safe_token(value: str) -> str:
    if not value or len(value) > 120 or any(
        character not in "abcdefghijklmnopqrstuvwxyz0123456789_.-" for character in value
    ):
        raise ValueError("operational log tokens must be bounded and safe")
    return value


def _safe_identifier(value: str) -> str:
    if not value or len(value) > 160 or any(
        not (character.isascii() and (character.isalnum() or character in "_.:-"))
        for character in value
    ):
        raise ValueError("operational log identifiers must be bounded and safe")
    return value
